google/cloud/storage/internal/xml_node.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/xml_node.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/absl_str_replace_quiet.h"
+#include "absl/strings/str_format.h"
+#include <stack>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+namespace {
+
+std::string EscapeXmlString(std::string const& val, bool for_text) {
+  auto ret =
+      absl::StrReplaceAll(val, {{"&", "&amp;"}, {"<", "&lt;"}, {">", "&gt;"}});
+  if (!for_text) {
+    ret = absl::StrReplaceAll(ret, {{"\"", "&quot;"}, {"'", "&apos;"}});
+  }
+  return ret;
+}
+
+}  // namespace
+
+std::string XmlNode::GetConcatenatedText() const {
+  if (!text_content_.empty()) {
+    // For non-tag element, just returns the text content.
+    return text_content_;
+  }
+  std::string ret;
+  std::stack<XmlNode const*> stack;
+  stack.push(this);
+  while (!stack.empty()) {
+    auto const* cur = stack.top();
+    stack.pop();
+    ret += cur->text_content_;
+    // put them into the stack in reverse order
+    for (auto it = cur->children_.rbegin(); it != cur->children_.rend(); ++it) {
+      stack.push(&*it);
+    }
+  }
+  return ret;
+}
+
+std::vector<XmlNode const*> XmlNode::GetChildren() const {
+  std::vector<XmlNode const*> ret;
```

Comment:
Consider:

```
std::vector<XmlNode const*> ret(children_.size());
std::transform(children_.begin(), children_.end(); ret.begin(), [](auto c) { return &c; });
return ret;
```

But if we use `std::shared_ptr<XmlNode>` that may become:

```
std::vector<std::shared_ptr<XmlNode const>> XmlNode::GetChildren() const {
  return {children_.begin(), children_.end()};
}
```

https://godbolt.org/z/eqeqW5KW4